### PR TITLE
Fix/10017 parse file pipe builder

### PR DESCRIPTION
**Pull request by thiagomini**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #10017

## What is the new behavior?

Now we can specify if a given `file` parameter is required in the `ParseFilePipeBuilder` or in the `ParseFilePipe`:
```ts
@UploadedFile(
  new ParseFilePipeBuilder()
    .addFileTypeValidator({
      fileType: 'json',
    })
    .build({
      fileIsRequired: false,
    }),
)
file?: Express.Multer.File,
``` 

With that, when a file is explicitly set to be optional, the ParseFileValidator won't throw an error. On the other hand, when the file is required (default behavior), it will now throw a 400 error.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: packages/common/test/pipes/file/parse-file.pipe.spec.ts [hunk 1]
```diff
@@ -116,5 +116,55 @@ describe('ParseFilePipe', () => {
         });
       });
     });
+
+    describe('when isFileOptional is true', () => {
```
Comment (thiagomini) [inline packages/common/test/pipes/file/parse-file.pipe.spec.ts:120]:
```suggestion
    describe('when fileIsOptional is true', () => {
```

File: packages/common/test/pipes/file/parse-file.pipe.spec.ts [hunk 2]
```diff
@@ -116,5 +116,55 @@ describe('ParseFilePipe', () => {
...
+      it('should pass validation if no file is provided', async () => {
+        const requestFile = undefined;
+
+        await expect(parseFilePipe.transform(requestFile)).to.eventually.eql(
+          requestFile,
+        );
+      });
+    });
+
+    describe('when isFileOptional is false', () => {
```
Comment (thiagomini) [inline packages/common/test/pipes/file/parse-file.pipe.spec.ts:137]:
```suggestion
    describe('when fileIsOptional is false', () => {
```

File: packages/common/test/pipes/file/parse-file.pipe.spec.ts [hunk 3]
```diff
@@ -116,5 +116,55 @@ describe('ParseFilePipe', () => {
...
+      it('should throw an error if no file is provided', async () => {
+        const requestFile = undefined;
+
+        await expect(parseFilePipe.transform(requestFile)).to.be.rejectedWith(
+          BadRequestException,
+        );
+      });
+    });
+
+    describe('when isFileOptional is not explicitly provided', () => {
```
Comment (thiagomini) [inline packages/common/test/pipes/file/parse-file.pipe.spec.ts:154]:
```suggestion
    describe('when fileIsOptional is not explicitly provided', () => {
```

File: sample/29-file-upload/e2e/app/app.e2e-spec.ts [hunk 4]
```diff
@@ -51,6 +51,19 @@ describe('E2E FileTest', () => {
       .expect(400);
   });
 
+  it('should throw when file is required but no file is uploaded', async () => {
+    return request(app.getHttpServer())
+      .post('/file/fail-validation')
+      .send()
```
Comment (thiagomini) [inline sample/29-file-upload/e2e/app/app.e2e-spec.ts:57]:
```suggestion
```

**Review by thiagomini — COMMENTED** [review]
Some minor fixes I'll work on soon

**Review by aribambang — APPROVED** [review]

File: packages/common/pipes/file/parse-file-options.interface.ts [hunk 5]
```diff
@@ -5,4 +5,9 @@ export interface ParseFileOptions {
   validators?: FileValidator[];
   errorHttpStatusCode?: ErrorHttpStatusCode;
   exceptionFactory?: (error: string) => any;
+
+  /**
+   * Defines if file parameter is optional. Default is false.
+   */
```
Comment (micalevisk) [inline packages/common/pipes/file/parse-file-options.interface.ts:11]:
```suggestion
   * Defines if file parameter is optional.
   * @default false
   */
```

File: packages/common/pipes/file/parse-file-options.interface.ts [hunk 6]
```diff
@@ -5,4 +5,9 @@ export interface ParseFileOptions {
...
+  fileIsOptional?: boolean;
```
Comment (micalevisk) [inline packages/common/pipes/file/parse-file-options.interface.ts:12]: what about

```suggestion
  required?: boolean;
```

and then the default would `true`.

File: packages/common/pipes/file/parse-file.pipe.ts [hunk 7]
```diff
@@ -25,13 +27,20 @@ export class ParseFilePipe implements PipeTransform<any> {
       exceptionFactory,
       errorHttpStatusCode = HttpStatus.BAD_REQUEST,
       validators = [],
+      fileIsRequired,
```
Comment (thiagomini) [inline packages/common/pipes/file/parse-file.pipe.ts:30]: changed to `required` semantic, but I maintained the full word `fileIsRequired` to be explicit here. In the consumer side, when creating a ParseFilePipe, I think it's worth it to specify what is required:
```ts
const parseFilePipe = new ParseFilePipe({
  validators: [],
  fileIsRequired: false,
});
```

File: packages/common/pipes/file/parse-file.pipe.ts [hunk 8]
```diff
@@ -25,13 +27,20 @@ export class ParseFilePipe implements PipeTransform<any> {
...
     this.exceptionFactory =
       exceptionFactory ||
       (error => new HttpErrorByCode[errorHttpStatusCode](error));
 
     this.validators = validators;
+
+    this.validators.unshift(
+      new RequiredFileValidator({
+        required: fileIsRequired,
+      }),
```
Comment (micalevisk) [inline packages/common/pipes/file/parse-file.pipe.ts:42]: isn't this redundant? not sure if I follow

```suggestion
      new RequiredFileValidator(),
```

File: packages/common/pipes/file/parse-file.pipe.ts [hunk 9]
```diff
@@ -25,13 +27,20 @@ export class ParseFilePipe implements PipeTransform<any> {
...
+    );
```
Comment (micalevisk) [inline packages/common/pipes/file/parse-file.pipe.ts:43]: can't we have just

```suggestion
    if (fileIsRequired)
      this.validators.unshift( new RequiredFileValidator() );
```

as the name of the validator is 'required'

Comment (thiagomini) [reply packages/common/pipes/file/parse-file.pipe.ts]: In the end, I had to completely remove the validator. The problem here was that this validator could validate an incoming optional file, but if there were any other validators expecting a file to exist, the error would be thrown again. In conclusion, I decided to go back to the previous implementation, only changing the interface to use the `isFileRequired` option.

**Review by micalevisk — APPROVED** [review]

**Comment by thiagomini** [comment]
@kamilmysliwiec waiting for your review here

**Comment by kamilmysliwiec** [comment]
lgtm

**Pull request merged**